static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,152 @@
+
+To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
+
+This was clever, but there are a few constraints we had to consider when designing a circuit around the FIFO:
+* The FIFO read and write setup and hold times had to be obeyed by the SPI clock and `CLK_VIS` signals
+* The SPI Clock could be in only divisions by 2 of 80 MHz (80, 40, 20, 10, etc)
+* If we emitted pixels too quickly, and VGA hardware could not read from the FIFO fast enough, we would blow the queue and drop pixels
+    * This could be mitigated by adding more space in the FIFO, but at a steep dollar cost
+* If we emitted pixels too slowly, and VGA hardware attempted to consume pixels faster than were being written then we'd also drop pixels
+* VGA hardware does not read from the FIFO at the pixel clock rate across the entire duration of the frame. Since the visible portion of the frame is only about 73% of the total time spent per frame (because of the blanking times), the average read rate per frame is actually less than the pixel clock rate
+
+The details in how we eneded up on our specific design will be elaberated on in another post, but suffice to say we had to coordinate design details across software, hardware, specific part choices, and part pricing.
+
+### Displaying Pixels
+To dispaly pixels, the `CLK_VIS` signal generated by the VGA timing hardware reads bytes from the FIFO and into a Programmable Logic Unit (PLU) which maps the 8 bits into 9 bits, 3 sets of 3 bits for each of the color channels - red, green, and blue. These 9 bits then get sent to the DAC, described in the following section.
+
+An important consideration for this stage is how long it takes to shuttle a byte from the FIFO into the DAC and finally into the monitor. Since there is processing of the read bytes it's possible that the pixels become late compared to `H_SYNC` and `V_SYNC` and this will show up as a horizontal shift. I.E. the "first" pixel could actually show up on the 2nd or 3rd pixel position on the screen, instead of the actual 1st pixel.
+
+{% include figure.html image="/assets/images/left-collum-timing-bug.jpg" caption="Notice the far left column of the screen in white next to the blue. It appears to wrap around from the far right side of the screen because the first pixel of the image is shifted over 2 or 3 positions" width=550 %}
+
+So a method to handle this kind of pixel shifting makes the picture quality better.
+
+### Digital to Analog Converter (DAC)
+Once the pixel byte has been taken out of the FIFO memory it's on its way to being displayed by the monitor. To be displayed this pixel byte must be converted into three analog voltages between 0 and 0.7 volts corresponding to the Red, Green, and Blue color channels. This is accomplished with a [R2R Resistor Ladder DAC]().
```

Comment:
"This is accomplished" -> "We accomplished this"—there are many and more ways to do DACs, and ours is cheap & easy & simple but certainly not exclusive of the other options.

Reply:
Also, linky?